apps/openmw/mwworld/actiontrap.cpp

```diff
@@ -1,40 +1,45 @@
 #include "actiontrap.hpp"
 
 #include "../mwmechanics/spellcasting.hpp"
+
 #include "../mwbase/environment.hpp"
 #include "../mwbase/world.hpp"
 
 namespace MWWorld
 {
-
-    void ActionTrap::executeImp(const Ptr &actor)
+    // actor activated object without telekinesis, so trap will hit
+    void ActionTrap::executeImp(const Ptr &actor) 
     {
         osg::Vec3f actorPosition(actor.getRefData().getPosition().asVec3());
-        osg::Vec3f trapPosition(mTrapSource.getRefData().getPosition().asVec3());
-        float activationDistance = MWBase::Environment::get().getWorld()->getMaxActivationDistance();
+    
+        MWMechanics::CastSpell cast(mTrapSource, actor);
+        cast.mHitPosition = actorPosition;
+        cast.cast(mSpellId);
 
-        // GUI calcs if object in activation distance include object and player geometry
-        const float fudgeFactor = 1.25f;
+        mTrapSource.getCellRef().setTrap("");
+    }
 
-        // Hack: if actor is beyond activation range, then assume actor is using telekinesis
-        // to open door/container.
-        // Note, can't just detonate the trap at the trapped object's location and use the blast
+    // actor activated object with telekinesis, so trap may or may not hit
+    void ActionTrap::executeImp(const Ptr &actor, float distance)
+    {
+        osg::Vec3f trapPosition(mTrapSource.getRefData().getPosition().asVec3());
+
+        // Note: can't just detonate the trap at the trapped object's location and use the blast
         // radius, because for most trap spells this is 1 foot, much less than the activation distance.
-        if ((trapPosition - actorPosition).length() < (activationDistance * fudgeFactor))
-        {
-            // assume actor touched trap
-            MWMechanics::CastSpell cast(mTrapSource, actor);
-            cast.mHitPosition = actorPosition;
-            cast.cast(mSpellId);
-        }
-        else
+        // In the original game, the range of traps varies according to the object. Tested objects showed between around 145 and 170.
+        // This gives a small margin where the player can be within activation range (192) and still use telekinesis to safely open a trapped object.
+        // This matters for teleport doors, which can't otherwise be activated with telekinesis. For now, using an average of 157.5 to
+        // approximate original game behavior.
+        float trapRange = 157.5;
+
+        if (distance < trapRange) // actor activated object within range of trap
+            executeImp(actor);
+        else // actor activated object outside range of trap
         {
-            // assume telekinesis used
             MWMechanics::CastSpell cast(mTrapSource, mTrapSource);
             cast.mHitPosition = trapPosition;
             cast.cast(mSpellId);
-        }
-        mTrapSource.getCellRef().setTrap("");
+            mTrapSource.getCellRef().setTrap("");
+        }   
     }
```

Comment:
Looks like there is some code duplication here. I would prefer a single method for both cases. You could treat a distance of -1 (or generally < 0) as "distance unavailable". Then we could have a single method executeImp(const MWWorld::Ptr& actor, float distance=-1).

As for whether or not telekinesis was used, you can check that inside the ```ActionTrap``` by observing the actor's magic effects. There is no need to call different methods of the ```Action``` depending on telekinesis or no telekinesis as the base ```Action``` itself has nothing to do with that. And it would be quite a mess if we added more states at some point (now you have 2 execute methods for telekinesis and no telekinesis, add another boolean state and you have 4, then 8, and so on).